tower-layer/src/layer_fn.rs

```diff
@@ -0,0 +1,23 @@
+use super::Layer;
+
+/// Returns a new `LayerFn` with the given closure.
+pub fn layer_fn<T>(f: T) -> LayerFn<T> {
```

Comment:
Note that there's no need for `LayerFn` to be a public type. This can return an anonymous type like:

```suggestion
pub fn layer_fn<S, T, F>(f: F) -> impl Layer<S, Service = T> + Clone
where
    F: Fn(S) -> T,
    F: Clone,
{
```

Reply:
True but that means users wont be able to name the type in type aliases which I find pretty convenient.

Reply:
The majority of cases can't be named, anyway, since you're usually passing in an `impl Fn(S) -> T`. You can only name the thing when it's a pure `fn(S) -> T`.

I don't think you _have to_ omit the `LayerFn` type; but having used a helper like this extensively in our codebase, we have never actually used the named type in practice.

Just a suggestion, though. Feel free to ignore ;)

Reply:
FWIW, I *think* (though I could be wrong) that it actually wouldn't be a breaking change to replace the `impl Layer` with a named type in the future if there's a need for it — since the `impl Layer` is unnameable, I don't think it would be possible for code to depend on the return type being an `impl Layer` rather than a named type, as long as it still implements `Layer`.

So, if we wanted to, we *could* merge this with `impl Layer` and add a named type if users actually need it. 

On the other hand, @davidpdrsn, if you *know* you have a use-case for being able to name the type, that's demand for it. :)

On the *third* hand, I believe `service_fn` returns a nameable type, so it might be better to be consistent...or change `service_fn` to return `impl Service`.

Reply:
I don't have a specific use case in mind for naming `LayerFn` right now and the use case I have for `ServiceFn` could technically be replaced with `impl Service`:

```rust
pub fn not_found<R>() -> ServiceFn<fn(R) -> Ready<Result<Response<Body>, Infallible>>> {
    // ...
}
```

I could also write

```rust
pub fn not_found<R>() -> impl tower::Service<
    R,
    Response = Response<Body>,
    Error = Infallible,
    Future = Ready<Result<Response<Body>, Infallible>>,
> + Clone {
    // ...
}
```

That is just a bit unwieldy so I went for `ServiceFn` but ultimately it doesn't matter.

So I don't feel strongly about it and I think you're right that changing `impl Layer` to `LayerFn` in the future wont be a breaking change 🤔 I guess if users really need to name the layer type they might as well define their own type and implement `Layer` for it. Don't think thats a big hassle.

I'll change the PR to use `impl Layer` instead.

Reply:
Changed in d86e41e375d65.